### PR TITLE
Add a separate thread for scoring

### PR DESCRIPTION
**Pull request by adamtorgerson**

### CODE REVIEW

File: liquidator/bam_scorer.h [hunk 1]
```diff
@@ -276,15 +398,31 @@ class BamScorer
...
             if (m_output)
             {
+                // TODO: if we want to run multiple score threads in parallel, we will need to
+                // add an output queue similar to input queue here
                 bam_write1(m_output, read);
             }
         }
     }
 
+    void fetch_helper(const bam1_t *read)
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove unused fetch_helper definition.

File: liquidator/bam_scorer.h [hunk 2]
```diff
@@ -276,15 +398,31 @@ class BamScorer
...
+        
+            m_queued_reads.push_back(raii_read);*/
+    }
+
     static int bam_fetch_func(const bam1_t* read, void* handle)
     {
         BamScorer& scorer = *static_cast<BamScorer*>(handle);
         scorer.score_read(read);
+        
+        //scorer.fetch_helper(read);
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove commented out code.

File: liquidator/bam_scorer.h [hunk 3]
```diff
@@ -81,14 +88,26 @@ class BamScorer
             std::cout << "#pattern name\tsequence name\tstart\tstop\tstrand\tscore\tp-value\tq-value\tmatched sequence" << std::endl;
         }
 
+        m_reading = true;
+        std::thread score_thread1(&BamScorer::score_thread_actual, this);
+        //std::thread score_thread2(&BamScorer::score_thread_actual, this);
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove commented out code.

File: liquidator/bam_scorer.h [hunk 4]
```diff
@@ -81,14 +88,26 @@ class BamScorer
...
             score_regions(region_file_path);
         }
         else
         {
+            printf("ALL READS\n");
             score_all_reads();
         }
+        m_reading = false;
+        score_thread1.join();
+        //score_thread2.join();
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove commented out code.

File: liquidator/bam_scorer.h [hunk 5]
```diff
@@ -187,13 +258,63 @@ class BamScorer
...
+            }
+            m_queue_mutex.unlock();
+            
+            while(vec_count < MAX_THREAD_CHUNK)
+            {
+                int ret;
+                ret = bam_read1(m_input, &(*vec)[vec_count].bam);
+                if(ret < 0)
+                {
+                    printf("EOF? %d\n", total_read);
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove debugging prints

File: liquidator/bam_scorer.h [hunk 6]
```diff
@@ -187,13 +258,63 @@ class BamScorer
...
+            m_queue_mutex.unlock();
+            
+            vec.reset(new std::vector < BamAllocator >);
+            vec_count = 0;
+            vec->resize(MAX_THREAD_CHUNK);
+            memset(&((*vec)[0]), 0, MAX_THREAD_CHUNK * sizeof(BamAllocator));
+
+            if(break_main)
+            {
+                printf("Breaking main loop\n");
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove debugging prints

File: liquidator/bam_scorer.h [hunk 7]
```diff
@@ -187,13 +258,63 @@ class BamScorer
...
+                break;
+            }            
         }
+
+        printf("vec total: %d\n", vec_total);
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove debugging prints

File: liquidator/bam_scorer.h [hunk 8]
```diff
@@ -81,14 +88,26 @@ class BamScorer
...
         }
 
+        m_reading = true;
+        std::thread score_thread1(&BamScorer::score_thread_actual, this);
+        //std::thread score_thread2(&BamScorer::score_thread_actual, this);
+        //std::thread score_thread3(&BamScorer::score_thread_actual, this);
+        //std::thread score_thread4(&BamScorer::score_thread_actual, this);
         if (!region_file_path.empty())
         {
+            printf("REGIONS\n");
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove debugging prints

File: liquidator/bam_scorer.h [hunk 9]
```diff
@@ -81,14 +88,26 @@ class BamScorer
...
             score_regions(region_file_path);
         }
         else
         {
+            printf("ALL READS\n");
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h]: Please remove debugging prints

File: liquidator/bam_scorer.h [hunk 10]
```diff
@@ -179,6 +198,58 @@ class BamScorer
...
+                score_read(&raii_read.bam);
+            }
+
+            //printf("scored: %d\n", scored_total);
+        }
+    }
+    
     void score_all_reads()
     {
         // todo: the unmapped reads seem to all be at the very end of the loop.
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h:244]: Realizing that I wrote this comment in 2015 is horrifying.

File: liquidator/bam_scorer.h [hunk 11]
```diff
@@ -179,6 +198,58 @@ class BamScorer
...
+    {
+        BamAllocator() { }
+
+        ~BamAllocator()
+        {
+            // note: we are not calling bam_destroy1(), which was just free()ing the
+            // bam and the bam data. since the vectors manage the bam memory,
+            // we just need to free the bam dat, but if samtools changes this will
+            // likely break
+            free(bam.data);
```
Comment (jdimatteo) [inline liquidator/bam_scorer.h:203]: This seems like a good trade off given that we are so concerned with performance.

**Comment by jdimatteo** [comment]
Looks good; I saw about a 30% speedup in my tests.

If you could remove the commented out code and debug print statements, that would be my preference.  Assuming you don't rebase, we can always look back at that commit to see the commented out code if we need it.  Feel free to add a switch (defaulting to off) to enable the debug style print statements if you'd prefer to leave them in.

**Comment by adamtorgerson** [comment]
Thanks, looking forward to discussing more tomorrow, some of the commented parts merit further discussion.

No problem removing the debugging stuff, it was honestly probably not ready for a pull but I got a little excited and this is my first time using this pull workflow.

**Pull request merged**